src/xb-value-bindings.c

```diff
@@ -0,0 +1,310 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ * vi:set noexpandtab tabstop=8 shiftwidth=8:
+ *
+ * Copyright (C) 2020 Endless OS Foundation LLC
+ *
+ * Author: Philip Withnall <withnall@endlessm.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN				"XbValueBindings"
+
+#include "config.h"
+
+#include <glib.h>
+
+#include "xb-opcode-private.h"
+#include "xb-value-bindings.h"
+
+typedef struct {
+	enum {
+		KIND_NONE,
+		KIND_TEXT,
+		KIND_INTEGER,
+	} kind;
+	union {
+		gchar *text;
+		guint32 integer;
+	};
+	GDestroyNotify destroy_func;
+} BoundValue;
+
+typedef struct {
+	/* Currently limited to two values since that’s all that any client
+	 * uses. This could be expanded to dynamically allow more in future. */
+	BoundValue values[2];
+	gpointer dummy[3];
+} RealValueBindings;
+
+G_STATIC_ASSERT (sizeof (XbValueBindings) == sizeof (RealValueBindings));
+G_STATIC_ASSERT (G_ALIGNOF (XbValueBindings) == G_ALIGNOF (RealValueBindings));
+
+G_DEFINE_BOXED_TYPE (XbValueBindings, xb_value_bindings,
+		     xb_value_bindings_copy, xb_value_bindings_free)
+
+/**
+ * xb_value_bindings_init:
+ * @self: an uninitialised #XbValueBindings to initialise
+ *
+ * Initialise a stack-allocated #XbValueBindings struct so it can be used.
+ *
+ * Stack-allocated #XbValueBindings instances should be freed once finished
+ * with, using xb_value_bindings_clear() (or `g_auto(XbValueBindings)`, which is
+ * equivalent).
+ *
+ * Since: 0.3.0
+ */
+void
+xb_value_bindings_init (XbValueBindings *self)
+{
+	RealValueBindings *_self = (RealValueBindings *) self;
+
+	_self->values[0].kind = KIND_NONE;
+	_self->values[1].kind = KIND_NONE;
```

Comment:
can we add a loop with a MAX_VALUES type thing? then we only have to change a constant at the top in the future.